File: src/main/java/com/example/demo/service/BikeService.java
package com.example.demo.service;

import org.springframework.stereotype.Service;
import org.springframework.web.bind.annotation.PathVariable;

import com.example.demo.dto.RegisterBicycleRequest;
import com.example.demo.model.Bicycle;
import com.example.demo.model.Owner;
import com.example.demo.repository.BikeRepository;
import com.example.demo.repository.OwnerRepository;

import jakarta.transaction.Transactional;

@Service
public class BikeService {
    private final BikeRepository bikeRepository;
    private final OwnerRepository ownerRepository;

    public BikeService(BikeRepository bikeRepository, OwnerRepository ownerRepository) {
        this.bikeRepository = bikeRepository;
        this.ownerRepository = ownerRepository;

    }
    
    @Transactional
    public Bicycle getBicycleBySerialNumber(String serialNumber) {
        return bikeRepository.findBySerialNumber(serialNumber).orElseThrow(() -> new RuntimeException("bike not found"));
    }
    @Transactional
    public Bicycle SetOwner(String serialNumber, String userEmail ){ 
        Bicycle bicycle = bikeRepository.findBySerialNumber(serialNumber).orElseThrow(() -> new RuntimeException("Bike not found"));
        Owner newOwner = ownerRepository.findByEmail(userEmail).orElseThrow(() -> new RuntimeException("Owner not found"));

        Owner oldOwner = bicycle.getOwner();
        
        if (oldOwner != null){
            oldOwner.removeBike(bicycle);
        }

        if (newOwner!=null){
            newOwner.addBicycle(bicycle);
        }
        return bicycle;
    }
    @Transactional
    public Bicycle registerBike(RegisterBicycleRequest bicycleRequest){
        Bicycle bicycle = new Bicycle();
        Owner owner = ownerRepository.findByEmail(bicycleRequest.getOwnerEmail()).orElseThrow(() -> new RuntimeException("could not find owner with to register bike on"));
        bicycle.setOwner(owner); 
        bicycle.setBrand(bicycleRequest.getBrand());
        bicycle.setModel(bicycleRequest.getModel());
        bicycle.setSerialNumber(bicycleRequest.getSerialNumber());
        bikeRepository.save(bicycle);
        return bicycle;
    }
}

File: src/main/java/com/example/demo/service/BikeService.kt
package com.example.demo.service

import com.example.demo.dto.RegisterBicycleRequest
import com.example.demo.model.Bicycle
import com.example.demo.repository.BikeRepository
import com.example.demo.repository.OwnerRepository
import jakarta.transaction.Transactional
import org.springframework.data.crossstore.ChangeSetPersister
import org.springframework.stereotype.Service
import java.util.function.Supplier

@Service
open class BikeService(private val bikeRepository: BikeRepository, private val ownerRepository: OwnerRepository) {
    @Transactional
    fun getBicycleBySerialNumber(serialNumber: String?): Bicycle {
        return bikeRepository.findBySerialNumber(serialNumber)
                ?: throw RuntimeException("bike not found")
    }

    @Transactional
    fun SetOwner(serialNumber: String?, userEmail: String?): Bicycle {
        val bicycle = bikeRepository.findBySerialNumber(serialNumber) ?: throw RuntimeException("bike not found")
        val newOwner = ownerRepository.findByEmail(userEmail)
        val oldOwner = bicycle.owner ?: throw RuntimeException("owner not found")
        oldOwner.removeBike(bicycle)
        newOwner.addBicycle(bicycle)
        return bicycle
    }

    @Transactional
    fun registerBike(bicycleRequest: RegisterBicycleRequest): Bicycle {
        val bicycle = Bicycle()
        val owner = ownerRepository.findByEmail(bicycleRequest.ownerEmail)
        bicycle.owner = owner
        bicycle.brand = bicycleRequest.brand
        bicycle.model = bicycleRequest.model
        bicycle.serialNumber = bicycleRequest.serialNumber
        bikeRepository.save(bicycle)
        return bicycle
    }
}